### DISCUSSION TITLE
Question regarding TimeSeriesForestClassifier, and Feature Creation from Data

### DISCUSSION THREAD

**Discussion by v4gadkari** [alan-turing-institute/sktime#3947] [Q&A]
Hello Everyone!

I am currently an undergraduate student who is tasked by my advisor to work on a time series classification problem. Upon looking for good resources, I stumbled across sktime, and I am very impressed! I have worked on some code, and have developed a few questions in regard to how I am using the estimators and the package in general. This is a multivariate time series problem. 

1. Questions regarding setting up training and testing sets for time series classification:
 - My training set consists of X_train: (variable_1, variable_2, variable_3, variable_4), y_train: (target_response)
 - variable_1-4 are each measurement taken at each timestep, so they are each time series themselves
 - Should I be including time in my training set? When I used the time series forest classifier, I could not get the feature importances, is this because I need the timestamp? Should I adjust my training set to include the time? 
 
2. Questions regarding Estimators, and preprocessing steps before using them
- Right now, I am feeding TimeSeriesForestClassifier (and other estimators) my raw data as it is, should I be "chunking" my data into intervals to create a series? 
- For example, instead of feeding my raw data which consists of the 4 individual series as my X_train, should I be creating a new dataset that consists of each feature being 5-second chunks of my data? ie. feature_1: (0sec-5sec of variable_1-4), feature_2: (5-10sec of variable 1_4) etc.? Then, do I pass this into the classifier? (5 sec is arbitrary for the sake of the example)

3. When I use an estimator, is the associated transformer being applied under the hood?
- for example, when I use Rocket, is the rocket transform already being applied? 
- When I use TimeSeriesForestClassifier, is it making these intervals before doing the ensemble learning?

**Comment by fkiraly** [comment]
Welcome to sktime, @v4gadkari!

To send this ahead, I think you are not using the term "variable" as we are, since you are saying that "variable 1" to "variable 4" are the same quantity or observable measured at four different points in time? So, in fact, there is just one variable, and each are measured at different time points.

If you aren't sure, it would help if you posted a dummy data set (e.g., code to create a data frame) to show. Not necessarily using the same variables.

1. if I understand correctly, you want to include an additional variable, which is the time stamp of the measurement? That would give each instance 2 variables (the original one, and time), observed at four time stamps, is this correct?

If so, then it may or may not help - you have to test, using a hold-out set (and ensure you have another hold-out set for after model development).

To do this quickly, you can use the `TimeSince` transformer in a `FeatureUnion` transformer pipeline.

2. I'm not sure whether I got how your data looks like or what you mean by chunks. Displaying an example might be helpful.

3. That depends on the estimator. The Rocket classifiers - as I think is detailed in the docstrings - are the same as a `ClassifierPipeline` with one of the various rocket transforms and then the `RidgeClassifierCV` from `sklearn`. So the answer is "yes" in that instance. You can build your own variants of transformer and classifier by using the `ClassifierPipeline`.

For the `TimeSeriesForestClassifier`, I think so, then it applies a decision tree, and then it does bagging (with sub-sampling).

**Reply by v4gadkari** [reply]
Thanks for the response! Here is an example of the way the data is set up. This is experimental data gathered from a few test scenarios. The goal is to predict the scenario a subject was apart of using their results. Each row corresponds to a measurement of a subject at a specific time. 

'Time': 0.014, 0.015, 0.016, ..... 18, 19, 20 

'Var1': 882.4, 883.5, 898.4,... 881.4, 838.6, 875.3

'Var2': 1.78, 1.65, 1.73, ..., 1.23, 1.45, 1.67

'Var3': 0, 0, 0.003, ..., 0.045, 0.342

'Var4': 111.2, 132.4, 0, ..., 0, 23.4, 123.4

'Var5': -1.76, -1.82, -1.45, ..., -1.34, 1.43

Response: (encoded as 0, 1, 2) (corresponding to the test scenario)

Subject: the subject number of the participant in the study

I want to find a function f, such that f(var1, var2, var3, var4, var5) -> Response, where the response is the test scenario. 

Hope that helps!
